src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -142,8 +144,28 @@ pub fn build_unit_dependencies<'a, 'gctx>(
     // which affect the determinism of the build itself. As a result be sure
     // that dependency lists are always sorted to ensure we've always got a
     // deterministic output.
-    for list in state.unit_dependencies.values_mut() {
-        list.sort();
+    // We do not sort build scripts to preserve the user-defined order.
+    for unit_dep in &mut state.unit_dependencies {
+        let is_multiple_build_scripts_enabled = unit_dep
+            .0
+            .pkg
+            .manifest()
+            .unstable_features()
+            .require(Feature::multiple_build_scripts())
+            .is_ok();
+
+        let list = unit_dep.1;
+        if is_multiple_build_scripts_enabled {
+            list.sort_by_key(|unit_dep| {
+                if unit_dep.unit.target.is_custom_build() {
+                    (0, None)
+                } else {
+                    (1, Some(unit_dep.clone()))
+                }
+            });
+        } else {
+            list.sort();
+        }
```

Comment:
`for (unit, list) in &mut state.unit_dependencies` can avoid accessing the tuple fields